crates/ruff_server/src/server.rs

```diff
@@ -0,0 +1,159 @@
+        let (conn, threads) = lsp::Connection::stdio();
+
+        let (id, params) = conn.initialize_start()?;
+
+        let init_params: types::InitializeParams = serde_json::from_value(params)?;
+
+        let client_capabilities = init_params.capabilities;
+        let server_capabilities = Self::server_capabilities(&client_capabilities);
+
+        let workspaces = init_params
+            .workspace_folders
+            .map(|folders| folders.into_iter().map(|folder| folder.uri).collect())
+            .or_else(|| init_params.root_uri.map(|u| vec![u]))
+            .ok_or_else(|| {
+                anyhow!("No workspace or root URI was given in the LSP initialization parameters. The server cannot start.")
+            })?;
+
+        let initialize_data = serde_json::json!({
+            "capabilities": server_capabilities,
+            "serverInfo": {
+                "name": crate::SERVER_NAME,
+                "version": crate::version()
+            }
+        });
+
+        conn.initialize_finish(id, initialize_data)?;
+
+        Ok(Self {
+            conn,
+            threads,
+            session: Session::new(&server_capabilities, &workspaces)?,
+        })
+    }
+
+    pub fn run(self) -> crate::Result<()> {
+        let result = main_loop_thread(move || Self::main_loop(&self.conn, self.session))?.join();
+        self.threads.join()?;
+        result
+    }
+
+    fn main_loop(connection: &Connection, session: Session) -> crate::Result<()> {
+        // TODO(jane): Make thread count configurable
+        let mut scheduler = schedule::Scheduler::new(session, 4, &connection.sender);
+        for msg in &connection.receiver {
+            let task = match msg {
+                lsp::Message::Request(req) => {
+                    if connection.handle_shutdown(&req)? {
+                        return Ok(());
+                    }
+                    api::request(req)
+                }
+                lsp::Message::Notification(notification) => api::notification(notification),
+                lsp::Message::Response(response) => {
+                    tracing::error!(
+                        "Expected request or notification, got response instead: {response:?}"
+                    );
+                    continue;
+                }
+            };
+            scheduler.dispatch(task);
+        }
+        Ok(())
+    }
+
+    fn server_capabilities(client_capabilities: &ClientCapabilities) -> types::ServerCapabilities {
+        let position_encoding = client_capabilities
+            .general
+            .as_ref()
+            .and_then(|general_capabilities| general_capabilities.position_encodings.as_ref())
+            .and_then(|encodings| {
+                encodings
+                    .iter()
+                    .cloned()
+                    .filter_map(|encoding| PositionEncoding::try_from(encoding).ok())
+                    .max() // this selects the highest priority position encoding
```

Comment:
Nit: It might be worth to implement `TryFrom` for `&LspEncoding` to avoid the cloning here (I don't know how expensive cloning the LSP encoding is but I don't see a reason why we can't implement `TryFrom` for a reference instead)